ext/cache/sqlite.rs

```diff
@@ -0,0 +1,401 @@
+                    response_status        INTEGER NOT NULL,
+                    response_status_text   TEXT,
+                    response_body_key      TEXT,
+                    FOREIGN KEY (cache_id) REFERENCES cache_storage(id) ON DELETE CASCADE,
+
+                    UNIQUE (cache_id, request_url)
+                )",
+          (),
+        )
+        .expect("failed to create request_response_list table");
+      SqliteBackedCache {
+        connection: Arc::new(Mutex::new(connection)),
+        cache_storage_dir,
+      }
+    }
+  }
+}
+
+#[async_trait]
+impl Cache for SqliteBackedCache {
+  /// Open a cache storage. Internally, this creates a row in the
+  /// sqlite db if the cache doesn't exist and returns the internal id
+  /// of the cache.
+  async fn storage_open(&self, cache_name: String) -> Result<i64, AnyError> {
+    let db = self.connection.clone();
+    tokio::task::spawn_blocking(move || {
+      let db = db.lock();
+      db.execute(
+        "INSERT OR IGNORE INTO cache_storage (cache_name) VALUES (?1)",
+        params![cache_name],
+      )?;
+      let cache_id = db.query_row(
+        "SELECT id FROM cache_storage WHERE cache_name = ?1",
+        params![cache_name],
+        |row| {
+          let id: i64 = row.get(0)?;
+          Ok(id)
+        },
+      )?;
+      Ok::<i64, AnyError>(cache_id)
+    })
+    .await?
+  }
+
+  /// Check if a cache with the provided name exists.
+  /// Note: this doesn't check the disk, it only checks the sqlite db.
+  async fn storage_has(&self, cache_name: String) -> Result<bool, AnyError> {
+    let db = self.connection.clone();
+    tokio::task::spawn_blocking(move || {
+      let db = db.lock();
+      let cache_exists = db.query_row(
+        "SELECT count(cache_name) FROM cache_storage WHERE cache_name = ?1",
+        params![cache_name],
+        |row| {
+          let count: i64 = row.get(0)?;
+          Ok(count > 0)
+        },
+      )?;
+      Ok::<bool, AnyError>(cache_exists)
+    })
+    .await?
+  }
+
+  /// Delete a cache storage. Internally, this deletes the row in the sqlite db.
+  async fn storage_delete(&self, cache_name: String) -> Result<bool, AnyError> {
+    let db = self.connection.clone();
+    let cache_storage_dir = self.cache_storage_dir.clone();
+    tokio::task::spawn_blocking(move || {
+      let db = db.lock();
+      let maybe_cache_id = db
+        .query_row(
+          "DELETE FROM cache_storage WHERE cache_name = ?1 RETURNING id",
+          params![cache_name],
+          |row| {
+            let id: i64 = row.get(0)?;
+            Ok(id)
+          },
+        )
+        .optional()?;
+      if let Some(cache_id) = maybe_cache_id {
+        let cache_dir = cache_storage_dir.join(cache_id.to_string());
+        if cache_dir.exists() {
+          std::fs::remove_dir_all(cache_dir)?;
+        }
+      }
+      Ok::<bool, AnyError>(maybe_cache_id.is_some())
+    })
+    .await?
+  }
+
+  async fn put(
```

Comment:
This `put` is racy. If a user inserts an item into the cache with a pending body, then attempts to read that item from the cache, the item read will either error (no such file), or will return partial body data.

We need to change how puts work:
1. When calling `cache.put()` you create a resource on disk for that cache, with an arbitrary name.
2. Then create a body writer resource containing the file handle, request / resp metadata, and a timestamp indicating when the `cache.put` was started.
3. The body is then written to the file.
4. When the body stream is closed, you insert the metadata into sqlite, _if_ the timestamp in the resource is larger than the last inserted timestamp for that URL in sqlite.

This prevents the race, as other callers will not be able to see the entry until the entire body data is written. Additionally this prevents a scenario where two `cache.put` operations are started in close proximity, but the one that is later finishes before the earlier one, and then the data from the later operation is lost.

Reply:
> When the body stream is closed, you insert the metadata into sqlite, if the timestamp in the resource is larger than the last inserted timestamp for that URL in sqlite.

What role does the timestamp play? I don't understand its role. Timestamp in `cache.put` will always be greater than the one in sqlite.

Reply:
added a timestamp to track